Virtual File system Switch (VFS) support

ramakrishnapallala opened the pull request:
The intention of these commits is
- To add the VFS like support in Zephyr with minimal changes in existing API
- Allow the system to have different file systems mounted simultaneously
- Decouple File Systems from applications and make them more like plug-in modules

The following things are work-in-progress: 
- Supporting multiple partitions or instances of same file system
- Test application which tests multiple file systems on multiple partitions

These commits are related to Issue #2978

Thank you.

Commit:
subsys: fs: Kconfig: Add log level support for fs subsystem

Add SYS_LOG_FS_LEVEL Kconfig option for file system subsystem.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
subsys: fs: Add Virtual File system Switch (VFS) support

Add support for Virtual File system Switch (VFS) by
introducing mount point concept to Zephyr. This allows
the applications to mount multiple file systems at
different mount points (ex: "/fatfs" and "/nffs"). The
mount point structure contains all the necessary info
required to instantiate, mount and operate on file system.

Decouple applications from directly accessing individual
file systems API's or internal functions by introducing
file system registration mechanism in VFS.

Move the file system defination and mount responsibility
to application so that application can decide which file system
to use and where to mount.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

carlescufi commented on subsys/fs/nffs_fs.c:
```diff
@@ -218,44 +222,44 @@ static int inode_to_dirent(struct nffs_inode_entry *inode,
 {
 	int rc;
 
 	k_mutex_lock(&nffs_lock, K_FOREVER);
```
If you're already locking at the VFS level, do you need to lock here? or do you expect those APIs to be directly callable even with VFS?

ramakrishnapallala replied:
@carlescufi , In VFS layer we are only protecting the operations on mount point liek adding new fs or mounting it.

The normal file and directory operations are not protected at VFS layer, these need to be taken care by individual file systems depending their implementation.
The scope of this commit is adapt the existing file system interfaces with VFS.

I expect more changes to nffs_fs/fat_fs in upcoming "multiple instances of the same file system" commits.

lpereira commented on subsys/fs/fs.c:
```diff
@@ -0,0 +1,529 @@
+	}
+	fs_path = &abs_path[match_len];
+
+	if (mp->fs->fs_ops->stat != NULL) {
```
Suggestion: this indirection can be reduced if `mp->fs` points to a structure with all the function pointers thrown in instead of having 3 separate structs for file, filesystem, and directory operations:

```C
if (mp->fs->stat != NULL) {
    ...
}
```

ramakrishnapallala replied:
Ok. In that case, single struct contains all the functions pointers to file, directory and file system operations.

lpereira reviewed (commented):
Patch LGTM; some minor nits, though.

lpereira commented on subsys/fs/fs.c:
```diff
@@ -0,0 +1,529 @@
+	int rc, match_len;
+
+	if ((strlen(file_name) <= 1) || (file_name[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
```
Newlines should not be added to SYS_LOG_ERR() calls.  (There's a configuration option to control whether newlines will be automatically added by syslog messages.)

ramakrishnapallala replied:
Agree. will fix it.

ramakrishnapallala commented on subsys/fs/fat_fs.c:
```diff
@@ -273,27 +279,69 @@ int fs_statvfs(struct fs_statvfs *stat)
-	ARG_UNUSED(dev);
-
-	res = f_mount(&fat_fs, "", 1);
+	printk("%s: start\n", __func__);
```
yes forgot to remove.

ramakrishnapallala commented on subsys/fs/fat_fs.c:
```diff
@@ -273,27 +279,69 @@ int fs_statvfs(struct fs_statvfs *stat)
 
 	__ASSERT((res == FR_OK), "FS init failed (%d)", translate_error(res));
 
+	printk("%s: end\n", __func__);
```
yes, will fix it.

nvlsianpu commented:
@ramakrishnapallala
I'm going to add API for rename files and directory (see API prototype below:), currently have it on my own fork.

Can you add suitable s_file_ops s_dir_ops entries?
I can commit nffs implementation for this API.

```
/**
 * @brief File o directory rename
 * 
 * Performs a rename and / or move of the specified source path to the
 * specified destination.  The source path can refer to either a file or a
 * directory.  All intermediate directories in the destination path must
 * already exist.  If the source path refers to a file, the destination path
 * must contain a full filename path, rather than just the new parent
 * directory.  If an object already exists at the specified destination path,
 * this function causes it to be unlinked prior to the rename (i.e., the
 * destination gets clobbered).
 *
 * @param from The source path.
 * @param to The destination path.
 *
 * @return 0 Success;
 * @retval -ERRNO errno code if error
 */
int fs_rename(const char *from, const char *to);
```

nvlsianpu commented on subsys/fs/nffs_fs.c:
```diff
@@ -537,4 +540,47 @@ static int fs_init(struct device *dev)
+	.fs_ops = &nffs_fs_ops,
+};
+
+static int fs_init(struct device *dev)
```
name: nffs_init, use it in SYS_INIT macro bellow

ramakrishnapallala replied:
Ok, will fix it.

ramakrishnapallala replied:
@nvlsianpu , I can add the fs_rename() API and corresponding VFS changes. You can add the nffs implementation. Is that sound good?

nvlsianpu commented:
>@nvlsianpu , I can add the fs_rename() API and corresponding VFS changes. You can add the nffs implementation. Is that sound good?
@ramakrishnapallala - yes, works for me.

rbtchc commented on subsys/fs/Kconfig:
```diff
@@ -12,6 +12,20 @@ config FILE_SYSTEM
+	prompt "File System log level"
+	depends on SYS_LOG
+	default 0
+	help
```
How about adding a range check "range 0 4" here?

ramakrishnapallala replied:
I just reused existing stuff, may we can do this at kernel level as one PR.

rbtchc commented on subsys/fs/fs.c:
```diff
@@ -0,0 +1,529 @@
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
```
SYS_LOG_ERR will append the newline if not explicitly disabled.
I suppose "\n" is unnecessary.

rbtchc commented on subsys/fs/fs.c:
```diff
@@ -0,0 +1,529 @@
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
```
Should check "abs_path != NULL" first before calling strlen()

ramakrishnapallala replied:
yes, I fixed it.

Commit:
tests: subsys: fs: Adapt fat_fs_api test as per VFS changes

Adapt fat_fs_api test as per the VFS layer changes.

Add new test for FatFs file system mount.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
tests: subsys: fs: Adapt nffs_fs_api test as per VFS changes

Adapt nffs_fs_api test as per the VFS layer changes.

Add new test for NFFS file system mount.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

nvlsianpu commented on subsys/fs/fs.c:
```diff
@@ -0,0 +1,529 @@
+	k_mutex_lock(&mutex, K_FOREVER);
+	/* check if requested file system is registered */
+	if (mp->type >= FS_TYPE_END ||  fs_map[mp->type] == NULL) {
+		SYS_LOG_ERR("requested FS not supported!!\n");
```
Change log message: Not supported sounds like not supported at all, while fs type might be just not enabled in configuration.

ramakrishnapallala replied:
Ok will change the log message.

ramakrishnapallala commented on subsys/fs/fs.c:
```diff
@@ -0,0 +1,529 @@
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(file_name) <= 1) || (file_name[0] != '/')) {
```
The idea is based on the mount point (/fatfs or/nffs) through these mount points only we are finding the mounted file systems and calling the right fs interface function. So when we create a file or open a file we need to give the absolute path.

nvlsianpu replied:
That I understood. Why a mount point must have `/` at beginning?

ramakrishnapallala replied:
This will help us draw the similarity to Unix like files and I was thinking we can have pseudo root file system on '/' in future based on need.

Also it will be more clear if an absolute path starts with '/' than simply a name, easy for reading and debugging purposes.

lpereira replied:
I would prefer for paths to begin with `/`; otherwise, it's going to be difficult to have a concept of "current working directory".

Commit:
subsys: fs: Kconfig changes to enable multiple file systems

Kconfig changes to enable or select multiple file systems.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
tests: subsys: fs: Add test for multiple file systems

Add a new test for testing multiple file systems mounted
simultaneously on Zephyr. The test enables FATFS and NFFS
togeather and uses RAM as backend storage device.

The intention of this test is to demonstrate multiple file systems
support and perform basic file and directory operations. This test
heavily reused the existing fat_fs_api and nffs_fs_api test sources.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@carlescufi @lpereira @nvlsianpu @rbtchc I have addressed your comments and feedback.

Please review the updated patches.

ramakrishnapallala commented:
@carlescufi @nvlsianpu @lpereira Please review the updated patches.

lpereira reviewed (approved):
I did not review the tests, but skimmed over the VFS layer and it LGTM.

It would be nice in the future to have a shell application to perform interactive filesystem tests.

ramakrishnapallala commented:
Sure @lpereira I will add the shell app while I push the changes for multiple partition support.

ramakrishnapallala commented:
Hi @carlescufi , Can you review the PR and share your feedback. Thanks.

ramakrishnapallala commented:
@carlescufi , Can you review the PR?

nashif commented on include/fs.h:
```diff
@@ -8,23 +8,40 @@
-#include <fs/fs_interface.h>
+#include <misc/dlist.h>
+
+#ifdef CONFIG_FAT_FILESYSTEM_ELM
```
can we avoid knowledge about the implementation of the various filesystems in the public header and do this on the filesystem implementation? This does not scale really well if we have to add this type of checks for every file system we add.

ramakrishnapallala replied:
@nashif  I have thought through this while implementing it. Right now we don't do much in our fs interface modules (subsys/fs/fat_fs.c) other than redirecting the calls to FS libs and the way we are handling things inside these modules are not uniform across file systems. I will relook into this once I am done with enabling multiple partition support for which I am working on disk susbsys.

For now, I will move the file system specific definitions info  fs_interface.h like the previous way and submit the PR. At least the public fs.h file will not have FS specific definitions.

nashif commented on include/fs.h:
```diff
@@ -39,13 +56,65 @@ enum fs_dir_entry_type {
  */
+struct fs_file_t {
+	union {
+#ifdef CONFIG_FAT_FILESYSTEM_ELM
```
same here.

nashif commented on tests/subsys/fs/multi-fs/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  test_multifs:
```
filesystem.multifs

ramakrishnapallala replied:
Ok will fix it.

nvlsianpu commented:
@ramakrishnapallala F. nffs_rename was lost after you had apply the last fix rebate - was this intentional?

nvlsianpu commented:
I will add it to https://github.com/zephyrproject-rtos/zephyr/pull/6408 - as it needs it.
